quinn-proto/src/connection/assembler.rs

```diff
@@ -44,51 +49,41 @@ impl Assembler {
             self.state = State::Unordered { recvd };
         }
 
-        Ok(self.pop().map(|(offset, data)| {
-            self.bytes_read += data.len() as u64;
-            (offset, data)
-        }))
-    }
-
-    // Get the the next ordered chunk
-    pub(crate) fn read(&mut self, max_length: usize) -> Result<Option<Bytes>, AssembleError> {
-        if self.is_stopped() {
-            return Err(AssembleError::UnknownStream);
-        } else if let State::Unordered { .. } = self.state {
-            return Err(AssembleError::IllegalOrderedRead);
-        }
-
         loop {
             let mut chunk = match self.data.peek_mut() {
                 Some(chunk) => chunk,
                 None => return Ok(None),
             };
 
-            if chunk.offset > self.bytes_read {
-                // Next chunk is after current read index
-                return Ok(None);
-            } else if (chunk.offset + chunk.bytes.len() as u64) <= self.bytes_read {
-                // Next chunk is useless as the read index is beyond its end
-                PeekMut::pop(chunk);
-                self.defragmented = self.defragmented.saturating_sub(1);
-                continue;
-            }
+            if ordered {
+                if chunk.offset > self.bytes_read {
+                    // Next chunk is after current read index
+                    return Ok(None);
+                } else if (chunk.offset + chunk.bytes.len() as u64) <= self.bytes_read {
+                    // Next chunk is useless as the read index is beyond its end
+                    PeekMut::pop(chunk);
+                    self.defragmented = self.defragmented.saturating_sub(1);
+                    continue;
+                }
 
-            // Determine `start` and `len` of the slice of useful data in chunk
-            let start = (self.bytes_read - chunk.offset) as usize;
-            if start > 0 {
-                chunk.bytes.advance(start);
-                chunk.offset += start as u64;
+                // Determine `start` and `len` of the slice of useful data in chunk
+                let start = (self.bytes_read - chunk.offset) as usize;
+                if start > 0 {
+                    chunk.bytes.advance(start);
+                    chunk.offset += start as u64;
+                }
             }
 
             return Ok(Some(if max_length < chunk.bytes.len() {
                 self.bytes_read += max_length as u64;
+                let offset = chunk.offset;
                 chunk.offset += max_length as u64;
-                chunk.bytes.split_to(max_length)
+                (offset, chunk.bytes.split_to(max_length))
             } else {
                 self.bytes_read += chunk.bytes.len() as u64;
                 self.defragmented = self.defragmented.saturating_sub(1);
-                PeekMut::pop(chunk).bytes
+                let chunk = PeekMut::pop(chunk);
+                (chunk.offset, chunk.bytes)
```

Comment:
We should probably respect `max_length` here too.

Reply:
What do you mean? This is the `else` branch for `if max_length < chunk_bytes.len()`, so we should already respect it, right?

Reply:
Whoops, confused this with the `if ordered` branch above. My bad.